### PR TITLE
Change Blockstore::get_slots_since() to return Vec instead of HashMap

### PR DESCRIPTION
**Pull request by steviez**
#### Problem
The only non-test caller of this function just iterates over the
elemnts, so need to incur extra cost of hashing.

#### Summary of Changes
Use Vec instead of HashMap.

Depends on https://github.com/solana-labs/solana/pull/27686 - the first commit is from that PR, so only need to review 2nd commit. Will rebase once that PR ships.

### CODE REVIEW

File: core/src/replay_stage.rs [hunk 1]
```diff
@@ -3367,15 +3367,11 @@ impl ReplayStage {
...
+            .expect("Expected Blockstore::get_slots_since() to succeed");
         generate_new_bank_forks_get_slots_since.stop();
 
         // Filter out what we've already seen
-        trace!("generate new forks {:?}", {
-            let mut next_slots = next_slots.iter().collect::<Vec<_>>();
-            next_slots.sort();
-            next_slots
-        });
+        trace!("generate new forks {:?}", next_slots.clone().sort());
```
Comment (steviez) [inline core/src/replay_stage.rs:3374]: They slots we pass to `get_slots_since()` are the keys from a `HashMap`. As such, they will be in no particular order. So, we do still need the sort in order to ensure the slots are in increasing order in this `trace!` statement; however, wondering if we can drop the `.sort()`, which would also allow dropping of the `.clone()`

File: ledger/src/blockstore.rs [hunk 2]
```diff
@@ -2971,20 +2971,19 @@ impl Blockstore {
...
     // Returns slots connecting to any element of the list `slots`.
-    pub fn get_slots_since(&self, slots: &[u64]) -> Result<HashMap<u64, Vec<u64>>> {
-        // Return error if there was a database error during lookup of any of the
-        // slot indexes
-        let slot_metas: Result<Vec<Option<SlotMeta>>> =
-            slots.iter().map(|slot| self.meta(*slot)).collect();
-
-        let slot_metas = slot_metas?;
-        let result: HashMap<u64, Vec<u64>> = slots
+    pub fn get_slots_since(&self, slots: &[u64]) -> Result<Vec<(u64, Vec<u64>)>> {
```
Comment (yhchiang-sol) [inline ledger/src/blockstore.rs:2974]: So this means we will have a public API change, can I know how we usually handle public API change?  Or the owners of validators will know the new release breaks their code and they just change accordingly?

Comment (steviez) [reply ledger/src/blockstore.rs:2974]: So this function is technically declared as `pub`; however, we don't strict API guidelines here similar to what we have with say the RPC methods.

Most validator operators use the software as-is, and as such, this signature change isn't even on their radar. For those that might be tweaking things, they are operating with the understanding that the codebase is still under heavy churn

**Comment by steviez** [comment]
I'm going to abandon this PR - I got off track and made the change before I realize how minimal the effect would be.

**Pull request closed without merging**